### ISSUE TITLE
Automate generation of the `SkyCoord` tab-completion documentation

### ISSUE THREAD

**Issue by eteq** [astropy/astropy#9321]
The documentation for `SkyCoord` has a section listing the attributes on a `SkyCoord` (http://docs.astropy.org/en/stable/coordinates/skycoord.html#attributes, which is in `docs/coordinates/skycoord.rst`).  This could/should be auto-generated instead of the current state of it having to be manually updated.

It looks to me like the tab-completion list is the same as you get from the following:
```
>>> from astropy.coordinates import SkyCoord
>>> sc = SkyCoord(1*u.deg, 2*u.deg)
>>> ["sc."+attrname for attrname in dir(sc) if not attrname.startswith('_')]
```
so in principle, it's just a matter of injecting that list into the documentation.  Would be nice to do this so that it's not necessary to manually keep that list up to date (or more likely: forget to do so...).

cc @adrn (who recently pointed out this table isn't up-to-date)

**Comment by tvarnish** [comment]
Hi @eteq @pllim. I'd like to try working on this issue as my first contribution. Would that be ok? 

I was thinking of generating the tab-completion list with a python script (as in your example), saving this list to a file, then using the ``.. include::`` Sphinx directive to include the output in the documentation. Does this sound like a reasonable solution, or am I missing something obvious?

**Comment by pllim** [comment]
Thank you for your interest, @tvarnish ! How about something that injects straight into the docstring, instead of having to generate a new file? Something like this:

https://github.com/astropy/astropy/blob/4b7fcafa0b3b4e30d10ce9ffe60aece4a6ef3675/astropy/units/utils.py#L78

**Comment by pllim** [comment]
p.s. I'll defer to @eteq for a final design decision as injecting into docstring is a little different from his original example.

**Comment by tvarnish** [comment]
Thanks @pllim - I’ll have a look into that, but I’ll wait to hear back from @eteq before starting a proper implementation.

**Comment by mshieldsdunn** [comment]
@eteq and @pllim - I am new to the package and interested in contributing and, as this issue seems to have gone a bit dormant, I tried to pick this up last night.

Did a bit of research and came across this interesting SO response to the question as to how to include the output of code into the docs (https://stackoverflow.com/questions/9809434/auto-generate-doctest-output-with-sphinx-extension).

TLDR is:

> "If you let Python produce the expected output, well.. it will no longer be expected (by the
> user/author), so the doctests will never fail, hence those tests will be useless."

Options I would propose (if you would be willing to consider) would be the following:

1. **Just "break the rule" in this particular case** - I think this is a bad engineering decision for a couple of reasons. Most importantly, I think this kind of case is *exactly* the kind of thing that the tests *should pick up*. Put differently, a maintainer shouldn't have to expend mental energy to notice that the docs are out of sync. The computer can do that and treating a mismatch as a failing test would provide that kind of alert. In the positive column, this should be fairly easy to implement as outlined by @eteq above. I do not suggest this approach.

2. **Write/use a custom sphinx extension** - This seems like a very heavyweight solution to a problem that isn't. More dependencies or additional code is just more (unnecessary) maintenance of something outside the core competency of the working group. Again, I do not suggest this approach.

3. **The @tvarnish approach** - This works nicely within the existing ecosystem without any additional dependencies, etc. My only thought is that sourcing an external file is a bit of unfortunate cognitive overhead. If, as an engineer, I am editing a `.rst` file, it would be preferable, all things considered, to have the entirety of the relevant context located within a single file. This is less than ideal, but I would have no issue approving a merge request using this approach.

4. **Remove the warning flag and let the test fail** - This seems like the smallest possible code change that achieves the semantically correct result. An engineer is saying, without the aid of the machine, "this is what I expect from this list comprehension...please tell me if that is not accurate." In any event this test breaks, it should (assuming something other than a **significant** re-factor), be a matter of observing the failing test in the documentation and then updating the hand-crafted output to match whatever corresponding changes were made in `SkyCoord`. Small code changes that are semantically indicative of their purpose seem like a pretty nice thing to me. 😄

**Please, please, please** take all the above with a large grain of salt. I am new-ish to Python and know nothing about the package or the underlying domain. This is just something that I am becoming interested in and, as a software engineer by trade, I thought this would be something fun to do (and hopefully helpful to others).

-m

**Comment by pllim** [comment]
@mshieldsdunn , I agree with your assessment. Thank you for the thoughtful comment! 

@eteq , as the person who suggested this originally, what say you?

**Comment by ludwigschwardt** [comment]
I had a quick peek at this as part of the inaugural [Astropy-SKA hackathon](https://github.com/orgs/ska-telescope/projects/1).

@mshieldsdunn To put your mind at ease, this is not a doctest per se that has to pass. It is really just a bit of documentation masquerading as code, which is why it also has a `# doctest: +SKIP` at the end of the `>>> sc.<TAB>` line. That, and you would not be able to inject a TAB keystroke anyway when running the doctest 😄 

My question is whether the tab completion example should resemble the reality of a recent IPython with `prompt_toolkit`, which looks more like this on the screen:
```Python
  >>> sc.<TAB>
         altaz                                cache
         apply_space_motion()                 cartesian
         barycentricmeanecliptic              cirs                                 >
         barycentrictrueecliptic              contained_by()
```

That would suggest just hardcoding the attributes like before, since now there is just the first 8 and not 115 :joy:

On the other hand, the documentation then mentions specific attributes by name, and it makes sense to have them appear in the above list.

**Comment by tactipus** [comment]
I'm interested in working on this issue. I'm looking through the thread and looking at the relevant files as I type.

**Comment by tvarnish** [comment]
Hi @tactipus! Feel free to pick this issue up; I've been too busy to work on it (and I didn't get very far with fixing it in the first place).

**Comment by AiIuru5** [comment]
Hello, I am new at this, but I found something that might help.

2. This extension allows python to be executed by sphinx on building HTML and the code to be displayed. The pertinent code would be stored in the rst file. It does mean we would have to add an extra dependency. https://pypi.org/project/sphinx_execute_code/


3.The original plan to have a generate docs definition in Ailuru5-main\astropy\coordinates\utils.py which gives a text file of code output(Ailuru5-main\docs\coordinates\TabCompleteList.txt)  which is ..inculded: in Ailuru5-main\docs\coordinates\skycoord.rst is a decent intermediate path. 

3.1 I have found that if  a make a Ailuru5-main\astropy\coordinates\utils.py like pllim suggested and have it call the routine in Ailuru5-main\astropy\coordinates\utils.py , it will run when "tox -e build_docs"  is called from Ailuru5-main\, but errors due to file permissions on windows because the repository is set so non-admins are read only. I'm not sure if this is intentional or a quirk of my machine/ how i setup the repository. I have left the ability to turn this on in Ailuru5-main\astropy\coordinates\utils.py

3.2 I don't love the idea of auto-injecting into Ailuru5-main\docs\coordinates\skycoord.rst because the contents of the document may need to be updated and i'd have to scan in the document to look for trigger words to put the string after. So for example look for "documentation:" and put the string after that, which would be susceptible to being inadvertently changed or appearing elsewhere in skycoords.rst compared to an include statement in sphinx. If there's a better way that's less brittle please let me know.

**Comment by eerovaher** [comment]
Tab-completion is a very useful feature, but it is not specific to `SkyCoord`, `astropy` or even to Python. Therefore I don't think `SkyCoord` documentation is the place where such basic computer literacy should be taught.

I also think that ensuring the documentation matches with what a user actually sees when they use tab-completion is completely hopeless. The output they see depends on whether they are using the Python (i.e. CPython) interpreter, IPython or an IDE. When using (at least) IPython only a limited number of completion options is displayed, so the documentation listing them all does not match what a user will see. Furthermore, the number of items that IPython displays can depend on the size of terminal screen and where on the screen is the prompt located.

If we want to display a list of all the public attributes of a `SkyCoord` instance in our documentation then I think we would be better off explicitly using the code snippet @eteq posted because at least then the result would not depend on what exact interpreter might be in use.

**Comment by clown1984** [comment]
Hello I would like to contribute. Is the issue still open? It seems this hasn't been solved, correct?

**Comment by AiIuru5** [comment]
It hasn't been solved. The community is in disagreement on how it should be
solved or if tab completion is necessary to teach in the docs.

The original solution wanted was clunky and ugly because it relied on the
python instance already being in the dev env, but it was supported by a
major contributor, so I tried to implement that and broke the checks for
the rest docs somehow (that's the linked pull request).

The nicer way to do it would be to install a sphinx dependency which would
have to be updated.

The best solution might be to leave it as a plain text example. So edit
skycords.rst to show some single case that's unlikely to be edited out of
the software being auto completed.

I'm willing to help you /work on this I just don't know what we're doing
here.

**Comment by eerovaher** [comment]
It would be best if @eteq as the author of this issue could clarify if it is still relevant.

**Comment by pllim** [comment]
I will just close this out as a pipe dream.
